Add flag for toggling permessage-deflate

code-asher opened the pull request:
This is a workaround for https://github.com/cdr/code-server/issues/3014 until we figure it out (if it is indeed related which it might not be; the zlib errors could merely be side effects of a connection issue).

jsjoeio commented:
> increase coverage by 0.12%
👏 You + Codecov helping us build a habit of increasing code coverage!

jsjoeio commented on src/node/cli.ts:
```diff
@@ -128,6 +129,7 @@ const options: Options<Required<Args>> = {
       "Disable update check. Without this flag, code-server checks every 6 hours against the latest github release and \n" +
       "then notifies you once every week that a new release is available.",
   },
+  enable: { type: "string[]" }, // There is no support for this flag.
```
❓ How come we say "There is no support for this flag?" is that to imply that it's not supported for code-server users?

code-asher replied:
Ah actually idk if we should keep this comment. I guess what I mean is actually "no guarantee these features will work".

jsjoeio replied:
Ahhh gotcha. Maybe change it to that instead? your call!

jawnsy replied:
Maybe we could change the "Enabling features" info message to say something like:

> Enabling the following experimental features:

and at the bottom, something like:

> The code-server project does not provide stability guarantees or commit to fixing bugs relating to these experimental features. When filing bug reports, please ensure that you can reproduce the bug with all experimental features turned off.

code-asher replied:
That's some premium copy :+1: Thanks! Will add.

jsjoeio commented on src/node/feature.ts:
```diff
@@ -0,0 +1,29 @@
+ *
+ * Return true if the feature was enabled or false if it's invalid.
+ */
+export function enableFeature(feature: string) {
+  for (const key in Feature) {
+    if (Feature[key as keyof typeof Feature] === feature) {
+      enabledFeatures.push(feature as Feature)
+      return true
+    }
+  }
+  return false
+}
+
+/**
+ * Return true if the feature is enabled. Features are always enabled in
+ * development mode to ensure we are testing them.
+ */
+export function isFeatureEnabled(feature: Feature) {
+  return enabledFeatures.includes(feature)
+}
```
This looks great and ++ for the documentation! 🎉 

nit: we may want to modify these functions to take `enabledFeatures` as an arg. That way, they're not relying and a global value, which could lead to unintended consequences later. Plus, it helps with testing.

code-asher replied:
Hmmm, yeah if we made `enabledFeatures`  immutable then we'd get purity.

Buuuut having to pass it around everywhere we need it seems a bit tiresome. :stuck_out_tongue: 

It also seems to create a bit of a mismatch. Feature flags are global settings but if you can pass them in it creates the illusion that they aren't. So the use and the design don't quite match up.

What kind of unintended consequences you were thinking of?

code-asher replied:
Actually...this is all entirely unnecessary lol. I'll just leave the enum then a caller can do `req.args.enable.includes(Feature.MyFeature)`.

jsjoeio reviewed (changes requested):
Overall, great work! 🎉 Just a couple nits about function being pure

code-asher commented:
Refactored; it's not any more pure but it's simpler. I think we could have an interesting discussion on whether we should pass `args` (and other variables) around instead of using a global access pattern at some point.

Coverage went down because of the code added in entry so I split some of those functions out in order to test them. One of them is now used for all the integration tests so those should reflect reality more closely now.

jsjoeio commented on test/unit/health.test.ts:
```diff
@@ -12,15 +12,15 @@ describe("health", () => {
 
   it("/healthz", async () => {
-    ;[, , codeServer] = await integration.setup(["--auth=none"], "")
+    codeServer = await integration.setup(["--auth=none"], "")
```
this looks way cleaner/easier to read 👏

jawnsy replied:
:+1: yea, the code before was :raised_eyebrow: kinda inscrutable, but then again, I'm not a JavaScript/TypeScript person

jsjoeio reviewed (approved):
Great work 👏

jsjoeio commented:
> I think we could have an interesting discussion on whether we should pass args (and other variables) around instead of using a global access pattern at some point

Agreed! I prefer passing args because it's easier to mock/test things. But global access does come in handy when writing code. It would be a good discussion though.

jsjoeio commented:
I'm going to fix the trivy scan in a PR soon so you can ignore for now
